File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def zigzagLevelOrder(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: List[List[int]]
        """
        if not root :
            return []
        queue =deque()
        queue.appendleft(root)
        result=[]
        k=0
        while queue:
            n = len(queue)
            level =[]
            for i in range(n): 
                top = queue.pop()
                level.append(top.val)
                if top.left :
                    queue.appendleft(top.left)
                if top.right :
                    queue.appendleft(top.right)
            if k%2 == 1 :
                level.reverse() 
            result.append(level)
            k +=1
        return result

File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        if(!root) return {};
        vector<vector<int>> res;
        queue<TreeNode*> q; 
        q.push(root);
        bool dir = 0;
        while(!q.empty()){
            auto size = q.size();
            vector<int> lvl;
            for(int i = 0 ; i < size ; i++){
                auto node = q.front();
                lvl.push_back(node->val);
                q.pop();
                if(node->left)
                    q.push(node->left);
                if(node->right)
                    q.push(node->right);

            }
            if(dir) reverse(lvl.begin(), lvl.end());
            dir = !dir;
            res.push_back(lvl);
        }
        return res; 
    }
};